### PR TITLE
add support for the Sign and Verify restful API

### PR DESCRIPTION
**Pull request by Liang1XZhao**
this restful API will exposed to public as:
- POST <ehsm_srv_address>/ehsm?Action=Sign
- POST <ehsm_srv_address>/ehsm?Action=Verify
add the corresponding test cases.
test: passed the unittest

Signed-off-by: Liang1XZhao <liang1x.zhao@intel.com>

### CODE REVIEW

File: ehsm_kms_service/test/test_kms_with_rest.py [hunk 1]
```diff
@@ -2,157 +2,146 @@
...
+
+appid = 202112101919
+nonce = ''
+nonce = nonce.join(random.choice('0123456789') for i in range(16))
+timestamp = int(time.time())
+
+
+params = {}
+params["appid"] = appid
+params["nonce"] = nonce
```
Comment (yang8621) [inline ehsm_kms_service/test/test_kms_with_rest.py:16]: nonce, timestamp, and signature should 1:1 mapping with each rest call.

File: ehsm_kms_service/test/test_kms_with_rest.py [hunk 2]
```diff
@@ -2,157 +2,146 @@
...
+params["timestamp"] = timestamp
+params["sign"] = 'timestamp'
```
Comment (yang8621) [inline ehsm_kms_service/test/test_kms_with_rest.py:18]: Hardcode a key, paired with appid, and calculate the HMAC-SHA256 as the signature.

File: ehsm_kms_service/ehsm_kms_params.js [hunk 3]
```diff
@@ -72,6 +72,48 @@ const ehsm_kms_params = {
     },
     aad,
   },
+  GenerateDataKeyWithoutPlaintext: {
+    cmk_base64: {
+      type: 'string',
+      maxLength: 1024,
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_params.js:78]: the cmk's max length is not 1024, it's decided by the keyspec, for now:
if the keyspec=AES128, then the length is 64,  if the keyspec=RSA3072, then the length seems is 2168

Comment (syan10) [reply ehsm_kms_service/ehsm_kms_params.js:78]: same issue for the above APIs define

File: ehsm_kms_service/ehsm_kms_server.js [hunk 4]
```diff
@@ -53,6 +53,10 @@ const _checkSign = function (req,res,next){
    const action = req.query.Action;
   const { payload } = req.body;
   const currentPayLoad = ehsm_kms_params[action];
+  if (!currentPayLoad) {
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_server.js:56]: indend issue

File: ehsm_kms_service/ehsm_kms_params.js (hunk 3 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/ehsm_kms_params.js:78]: it will be fix

File: ehsm_kms_service/test/test_kms_with_rest.py (hunk 2 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/test/test_kms_with_rest.py:18]: it will be fix

**Comment by yang8621** [comment]
BTW, the encryption/decryption of RSA3072 will also be added in the next patches, right?

File: core/App/ehsm_napi.h [hunk 5]
```diff
@@ -82,20 +93,41 @@ typedef struct {
...
  
-  static char* readData(char* jsonChar, std::string key){
+    static char* readData_string(char* jsonChar, std::string key){
         std::string retVal;
         std::string jsonString = jsonChar;
         key = "\"" + key + "\"";
 
-        int startIndex = jsonString.find(key) + key.size() + 1;
-        std::string subStr = jsonString.substr(startIndex);
+        int resultIndex = jsonString.find("\"result\"") + 8;
```
Comment (yang8621) [inline core/App/ehsm_napi.h:101]: What does 8 mean?  Could add some comments or use sizeof("xxxx")/strlen("xxx") instead?

File: ehsm_kms_service/ehsm_kms_params.js (hunk 3 above)
Comment (syan10) [reply ehsm_kms_service/ehsm_kms_params.js:78]: you can just set the hardcode max length for now(e.g. 4K, don't set the same length with plaintex data), maybe later we can refine the cmk's parameter to the cmk's ID which is a fix length.

File: core/App/ehsm_napi.h (hunk 5 above)
Comment (Liang1XZhao) [reply core/App/ehsm_napi.h:101]: It will be fix.

File: ehsm_kms_service/test/test_kms_with_rest.py [hunk 6]
```diff
@@ -2,157 +2,146 @@
 import json
 import argparse
 import base64
+import time
+import random
+
+appid = 202112101919
+nonce = ''
+nonce = nonce.join(random.choice('0123456789') for i in range(16))
```
Comment (yang8621) [inline ehsm_kms_service/test/test_kms_with_rest.py:10]: It's ranged from 0~0x2386F26FC0FFFF, 6 bytes~7 bytes.

I checked the design doc, where the nonce length is defined as 16. Does the 16 mean 16 Bytes?
If yes, nonce = ''.join(random.choice('0123456789ABCDEF') for i in range(32))
I think 4 bytes of the length is enough: nonce = base64(os.urandom(4))

Comment (yang8621) [reply ehsm_kms_service/test/test_kms_with_rest.py:10]: int.from_bytes(os.urandom(4), "big")

Comment (Liang1XZhao) [reply ehsm_kms_service/test/test_kms_with_rest.py:10]: It will use int.from_bytes(os.urandom(16), "big")

**Comment by syan10** [comment]
> BTW, the encryption/decryption of RSA3072 will also be added in the next patches, right?

**Comment by Liang1XZhao** [comment]
> BTW, the encryption/decryption of RSA3072 will also be added in the next patches, right?


> > BTW, the encryption/decryption of RSA3072 will also be added in the next patches, right?

yes

**Pull request closed without merging**